File: example-apps/javademoapp/src/main/java/com/stytch/javademoapp/MainViewModel.java
package com.stytch.javademoapp;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.stytch.sdk.common.StytchObject;
import com.stytch.sdk.common.StytchResult;
import com.stytch.sdk.common.network.models.BasicData;
import com.stytch.sdk.consumer.StytchClient;
import com.stytch.sdk.consumer.network.models.SessionData;
import com.stytch.sdk.consumer.network.models.UserData;
import com.stytch.sdk.consumer.sessions.Sessions;

import kotlin.Unit;

public class MainViewModel extends ViewModel {
    private MutableLiveData<StytchState> stytchState;
    public MutableLiveData<StytchState> getStytchState() {
        if (stytchState == null) {
            stytchState = new MutableLiveData<>(StytchState.DEFAULT);
        }
        return stytchState;
    }


    Unit handleInitializationChange(Boolean isInitialized) {
        StytchState newState = new StytchState(
            isInitialized,
            StytchClient.getSessions().getSync(),
            StytchClient.getUser().getSyncUser()
        );
        getStytchState().postValue(newState);
        if (isInitialized) {
            StytchClient.getUser().onChange(this::handleUserChange);
            StytchClient.getSessions().onChange(this::handleSessionChange);
        }
        return Unit.INSTANCE;
    }

    private Unit handleUserChange(StytchObject<UserData> stytchUser) {
        UserData userData = null;
        if (stytchUser instanceof StytchObject.Available) {
            userData = ((StytchObject.Available<UserData>) stytchUser).getValue();
        }
        StytchState newState = new StytchState(
            true,
            StytchClient.getSessions().getSync(),
            userData
        );
        getStytchState().postValue(newState);
        return Unit.INSTANCE;
    }

    private Unit handleSessionChange(StytchObject<SessionData> stytchSession) {
        SessionData sessionData = null;
        if (stytchSession instanceof StytchObject.Available) {
            sessionData = ((StytchObject.Available<SessionData>) stytchSession).getValue();
        }
        StytchState newState = new StytchState(
            true,
            sessionData,
            StytchClient.getUser().getSyncUser()
        );
        getStytchState().postValue(newState);
        return Unit.INSTANCE;
    }

    void logout() {
        StytchClient.getSessions().revoke(new Sessions.RevokeParams(), (StytchResult<BasicData> result) -> Unit.INSTANCE);
    }
}

File: example-apps/stytchexampleapp/src/main/java/com/stytch/stytchexampleapp/MainViewModel.kt
package com.stytch.stytchexampleapp

import android.os.Parcelable
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.stytch.sdk.common.StytchObjectInfo
import com.stytch.sdk.consumer.StytchClient
import com.stytch.sdk.consumer.network.models.SessionData
import com.stytch.sdk.consumer.network.models.UserData
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.WhileSubscribed
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.parcelize.Parcelize

class MainViewModel : ViewModel() {
    var authenticationState: StateFlow<AuthenticationState> =
        combine(
            StytchClient.isInitialized,
            StytchClient.user.onChange,
            StytchClient.sessions.onChange,
        ) { isInitialized, stytchUser, stytchSession ->
            val userData =
                if (stytchUser is StytchObjectInfo.Available) {
                    stytchUser.value
                } else {
                    null
                }
            val sessionData =
                if (stytchSession is StytchObjectInfo.Available) {
                    stytchSession.value
                } else {
                    null
                }
            AuthenticationState(
                isInitialized = isInitialized,
                userData = userData,
                sessionData = sessionData,
            )
        }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), AuthenticationState())
}

@Parcelize
data class AuthenticationState(
    val isInitialized: Boolean = false,
    val userData: UserData? = null,
    val sessionData: SessionData? = null,
) : Parcelable
